### ISSUE TITLE
Is it possible to configure the schema to allow a relationship with multiple types?

### ISSUE THREAD

**Issue by paul-weissman** [aribouius/jsonapi-react#9]
Is there scheme syntax for configuring a relationship that allows an item to be related to a number of different types?

e.g. something like this:
```
const schema = {
  page: {
    type: 'page',
    relationships: {
      media: {
        type: ['image', 'video],
      }
    }
  },
  image: {
    ...
  },
  video: {
    ...
  }
}
```

**Comment by aribouius** [comment]
Hey @paul-weissman would you mind expanding a bit on what your data model looks like, and what specific issue you are running into.

I.e. is your `media` resource polymorphic, or is it more along the lines of a single-table inheritance model?

**Comment by aribouius** [comment]
@paul-weissman it would also be useful to see what your API response looks like, if possible.

**Comment by paul-weissman** [comment]
Hi @aribouius, thanks for responding.

Firstly, the good news... I've found that with the schema set as per my example at the top of the thread does seem to work, at least it doesn't break and my relationships still get nested inside the parent objects. I wonder if it's then not doing things that it's supposed to do such as perhaps removing attributes that aren't referenced in the schema (does it do that normally?).

The only issue I have now is that when I receive the normalized version, I can't see what type each related entity has. I could try to extrapolate it in code based on the attributes and relationships present but that would be horribly brittle. For now I have patched this locally to retain the "type" property on all elements and have raised a PR (https://github.com/aribouius/jsonapi-react/pull/10).

To answer your question, an example use case is any web page that is constructed in "slices" which are each a different thing, e.g. a text block, a video, a callout, an embedded testimonial or Tweet, etc. Each of those things has a different data structure but are all sequental values of a property of the page.

Here is an example response (I've cleaned out links and things for brevity):

```
{
    "data": {
        "type": "page",
        "id": "1",
        "attributes": {
            "title": "Example page",
            "body": "Body copy here"
        },
        "relationships": {
            "field_content_sections": {
                "data": [
                    {
                        "type": "rich_text",
                        "id": "2"
                    },
                    {
                        "type": "video",
                        "id": "3"
                    },
                    {
                        "type": "callout",
                        "id": "4"
                    },
                    /* etc */
                ]
            }
        }
    },
    "included": [
        {
            "type": "rich_text",
            "id": "2",
            "attributes": {
                "text": "Text goes here",
            }
        },
        {
            "type": "video",
            "id": "3",
            "attributes": {
                "video_path": "/videos/video.file",
            }

        },
        {
            "type": "callout",
            "id": "4",
            "attributes": {
                "text": "Text goes here",
                "image": "/images/image.file",
                "link": "http://example.com"
            }

        },
        /* etc */
    ]
}
```

**Comment by aribouius** [comment]
@paul-weissman hey thanks for sharing that information, and for opening up that PR, I always appreciate contributions!

The way you have your response modeled is pretty interesting.  Having multiple resource types underneath a single relationship type like that is a bit edge-case, but doesn't seems to be outside the JSON-API spec, so it's certainly something I want to support in some capacity.

My main concern with the PR you opened is that it combines JSON-API schema information (the `type` field) with the resource's attributes, which is problematic for any resource that actually has a `type` field.  As such it would be a breaking change, and problematic for existing users who do have models with a `type` field (I'm one of them). 

I would personally approach this problem from one of two ways...

**1)** While not highlighted well in the docs, the `resolve` functionality provided by the [schema](https://github.com/aribouius/jsonapi-react#schema-definition) can also be used to append new fields.  The canonical example for why you might use it would be the following example:
```javascript
new ApiClient({
  schema: {
    user: {
      fields: {
        full_name: (_, attrs) => {
          return attrs.first_name + ' ' + attrs.last_name
        }
      }
    }
  },
})
```

You could utilize this to simply append your `type` field to whichever resource necessary, which would yield the same result as the PR you opened. 
```javascript
new ApiClient({
  schema: {
    rich_text: {
      fields: {
        type: () => 'rich_text'
      }
    }
  },
})
```

**2)** Reconsider how your data is modeled.  While I don't think the way your current data structure is modeled is _wrong_ per-se, the fact that stripping away the JSON-API schema metadata presents a problem might be a hint it could be improved.  For example, you are presenting both images and videos as a "media" type, yet they don't seem to share a common interface, or offer any way to differentiate between them, without the assistance of the JSON-API schema.   Any of the following approaches would likely provide beneficial decoupling from the JSON-API schema:
- Make all media types implement a common interface, that allows the client to handle each one properly without the need to differentiate between them. 
- Include an actual `type` field to each media instance, or something similar that allows for differentiation.
- Include the different media types as separate fields.

Hope this helps!

**Comment by paul-weissman** [comment]
Thanks @aribouius for that comprehensive response.

I couldn't get your example with the field callback working directly, but I assume you were referring using to the `resolve` property. I've implemented this and it works.

Incidentally, I may create another pull request changing `attrs[field] = ref.resolve(attrs[field], attrs)` to `attrs[field] = ref.resolve(attrs[field], attrs, rec)`, I don't think adding the third argument would cause any BC issues as current usages expecting two argument would continue to work as before.

This would however allow user to pull data from for instance the `links` property. I have a use case  where third party code injects derivative paths for image entities into the links object.

Let me know if you'd consider this addition and if so I'll raise a PR.

On your point 2, thanks for the observations. To be clear, I actually removed the JSON-API schema metadata in my example for the sake of brevity. I do legitimately feel that the data model we have in place is more optimal for our and probably many other project's IA. If one embeddable thing has for example a color selection field and another doesn't then it makes little sense to have storage or a an API representation for a color on the second thing. In our case, each of the different things than can be put together to make up the sequential sections of a page have very little in common and it makes sense to maintain optimal data structures for each.

**Comment by paul-weissman** [comment]
Hi @aribouius. Thanks again for your input on this.

I've submitted a PR (https://github.com/aribouius/jsonapi-react/pull/11). I didn't update the README or tests as they don't currently cover the second argument to the resolve() function in any detail anyway.

No worries about sharing your thoughts on the data model by the way... I welcome feedback on things like that. That's how we grow as devs after all.

**Comment by aribouius** [comment]
@paul-weissman your change is now available on version `0.0.16`.

Thanks!
